staging/src/k8s.io/apiserver/pkg/endpoints/filters/metrics.go

```diff
@@ -68,15 +69,69 @@ var (
 		},
 		[]string{"result"},
 	)
+
+	authorizationAttemptsCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Name:           "authorization_attempts_total",
+			Help:           "Counter of authorized attempts.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"result"},
+	)
+
+	authorizationLatency = metrics.NewHistogramVec(
+		&metrics.HistogramOpts{
+			Name:           "authorization_duration_seconds",
+			Help:           "Authorization duration in seconds broken out by result.",
+			Buckets:        metrics.ExponentialBuckets(0.001, 2, 15),
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"result"},
+	)
+
+	authorizationDecisionAnnotationCounter = metrics.NewCounterVec(
+		&metrics.CounterOpts{
+			Name:           "authorization_decision_annotations_total",
+			Help:           "Annotation values set in advanced audit.",
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"decision"},
+	)
 )
 
 func init() {
 	legacyregistry.MustRegister(authenticatedUserCounter)
 	legacyregistry.MustRegister(authenticatedAttemptsCounter)
 	legacyregistry.MustRegister(authenticationLatency)
+	legacyregistry.MustRegister(authorizationAttemptsCounter)
+	legacyregistry.MustRegister(authorizationLatency)
+	legacyregistry.MustRegister(authorizationDecisionAnnotationCounter)
+}
+
+func recordAuthorizationMetrics(ctx context.Context, authorized authorizer.Decision, err error, authStart time.Time, authFinish time.Time) {
+	var resultLabel, decisionAnnotationLabel string
+
+	switch {
+	//Follow the behavior in withAuthorization, even encounter evaluation errors and still allow the request
+	//Will not judge whether there is an error
+	case authorized == authorizer.DecisionAllow:
+		decisionAnnotationLabel = decisionAllow
+	case err != nil:
+		resultLabel = errorLabel
+	case authorized == authorizer.DecisionDeny:
+		decisionAnnotationLabel = decisionForbid
+		resultLabel = successLabel
+	case authorized == authorizer.DecisionNoOpinion:
+		// continue to the next authorizer
+		resultLabel = successLabel
```

Comment:
It will be quite confusing to return a success when the authorizer didn't make a decision. I would understand a result="success" label of authorization_attempts_total to represent an authorization, which is not the case with DecisionNoOpinion.

Maybe we could just have a `decision` label instead of `result`?

Reply:
> It will be quite confusing to return a success when the authorizer didn't make a decision. I would understand a result="success" label of authorization_attempts_total to represent an authorization, which is not the case with DecisionNoOpinion.
> 
> Maybe we could just have a `decision` label instead of `result`?

In the Authorize interface, authorizer actions and errors are not mutually exclusive; an authorizer like RBAC may encounter an evaluation error and still allow the request.
I don't think it is possible to simply use the `decision` label instead of `result`.
metrics should be divided into the following cases in order
1. when action is allow, the error is ignored and the count of metrics with `decision="allow"` and `result="success"` is added to 1
2. when error is non-empty, the metrics count of `{result="error"}` is added to 1
3. When action is forbid, also ignore the error and add 1 to the metrics count of `decision="fobid` and `result="success"`.
4. When action is NoOpinion, add 1 to the metrics count of `{decision="no opinion"}`.

The above is just a shallow understanding after reading the code, do you think this is reasonable?
Please review my latest commit code, it follows the above logic

Reply:
Thank you for your explanation, I agree with you and have used the first suggestion to make changes, please review code